partners_api/yandex_api.cpp

```diff
@@ -0,0 +1,159 @@
+#include "partners_api/yandex_api.hpp"
+
+#include "platform/http_client.hpp"
+
+#include "geometry/latlon.hpp"
+
+#include "base/logging.hpp"
+#include "base/thread.hpp"
+
+#include "3party/jansson/myjansson.hpp"
+
+#include "private.h"
+
+#include <iomanip>
+#include <limits>
+#include <sstream>
+
+namespace
+{
+std::string const kTaxiInfoUrl = "https://taxi-routeinfo.taxi.yandex.net";
+std::string g_yandexUrlForTesting = "";
+
+bool RunSimpleHttpRequest(string const & url, string & result)
+{
+  platform::HttpClient request(url);
+  request.SetRawHeader("Accept", "application/json");
+  if (request.RunHttpRequest() && !request.WasRedirected() && request.ErrorCode() == 200)
+  {
+    result = request.ServerResponse();
+    return true;
+  }
+
+  return false;
+}
+
+bool CheckYandexAnswer(json_t const * answer)
+{
+  if (answer == nullptr)
+    return false;
+
+  if (!json_is_array(answer))
+    return false;
+
+  if (json_array_size(answer) <= 0)
+    return false;
+
+  return true;
+}
+
+string GetYandexURL()
+{
+  if (!g_yandexUrlForTesting.empty())
+    return g_yandexUrlForTesting;
+
+  return kTaxiInfoUrl;
+}
+}  // namespace
+
+namespace taxi
+{
+namespace yandex
+{
+bool RawApi::GetTaxiInfo(ms::LatLon const & from, ms::LatLon const & to, std::string & result)
+{
+  std::stringstream url;
+  url << std::fixed << std::setprecision(6) << GetYandexURL()
+      << "/taxi_info?clid=" << YANDEX_CLIENT_ID << "&apikey=" << YANDEX_API_KEY
+      << "&rll=" << from.lon << "," << from.lat << "~" << to.lon << "," << to.lat
+      << "&class=econom,business,comfortplus,minivan,vip";
+
+  return RunSimpleHttpRequest(url.str(), result);
+}
+
+/// Requests list of available products from Yandex. Returns request identificator immediately.
+void Api::GetAvailableProducts(ms::LatLon const & from, ms::LatLon const & to,
+                               ProductsCallback const & successFn,
+                               ErrorProviderCallback const & errorFn)
+{
+  threads::SimpleThread([from, to, successFn, errorFn]()
```

Comment:
Why bad:
* lifetime of these threads is unknown
* number of such threads is unlimited